feat: add incremental_overlap_seconds parameter for watermark safety margin

ZdenekSrotyr opened the pull request:
# Add incremental_overlap_seconds parameter

Fixes SUPPORT-13828

## Problem
Records modified near the watermark boundary may be missed between incremental runs due to clock skew or delayed commits.

## Solution
New optional parameter `incremental_overlap_seconds` (default: 0) subtracts specified seconds from the last watermark to ensure records near the boundary are re-fetched.

**Example:**
- Last watermark: `2025-10-16T10:00:00.000Z`
- With overlap 60s: query uses `2025-10-16T09:59:00.000Z`
- Duplicates are auto-deduplicated by primary key

## Changes
- Added `incremental_overlap_seconds` to config schema
- Modified watermark logic in `_build_soql_query()`
- Updated README with documentation

**Backward compatible:** Default value 0 preserves existing behavior.

---
Link to Devin run: https://app.devin.ai/sessions/6f20e942b9004c6b8b2972658da68f5e
Requested by: @ZdenekSrotyr

Impact analysis:
https://keboola.cloud.gooddata.com/dashboards/#/workspace/component_factory/dashboard/a65e1491-4933-4dde-8252-4b8180b63466

devin-ai-integration[bot] committed:
feat: add incremental_overlap_seconds parameter for watermark safety margin

- Add incremental_overlap_seconds configuration parameter to prevent missing records
- Subtract specified seconds from last watermark when querying
- Default value is 0 (backward compatible - no change to existing behavior)
- Add documentation in README.md with usage example
- Log overlap application for transparency

Co-Authored-By: zdenek.srotyr@keboola.com <zdenek.srotyr@keboola.com>

soustruh commented on src/component.py:
```diff
@@ -492,7 +494,20 @@ def _build_soql_query(salesforce_client: SalesforceClient, params: dict, last_ru
         if incremental and incremental_fetch and incremental_field and last_run:
-            soql_query.set_query_to_incremental(incremental_field, last_run)
+            adjusted_last_run = last_run
+            if incremental_overlap_seconds and incremental_overlap_seconds > 0:
```
I see just a minor discrepancy here. Either:
- We trust users they will edit this value only using the UI, so the integer will always be positive and the check `incremental_overlap_seconds > 0` is redundant, or
- We want to protect them against setting an invalid (negative) value for example in the debug mode, but in such case, we should warn them about an invalid value, not fail (correct me if I'm wrong, but setting the value to `-1` would cause a `NameError` as `adjusted_last_run` in the `soql_query.set_query_to_incremental(incremental_field, adjusted_last_run)` call is not defined, huh?)

soustruh replied:
Oh never mind, I was didn't see it correctly here in the web interface… It won't crash, but it also won't inform user about their value being invalid 🙂

soustruh reviewed (changes requested):
Please check my comment, I think an invalid value in configuration can cause the component to crash… 🤔

devin-ai-integration[bot] committed:
fix: add warning for negative incremental_overlap_seconds values

- Explicitly check for negative values and log a warning
- Improve code clarity by separating validation logic
- Address review feedback from @soustruh

Co-Authored-By: zdenek.srotyr@keboola.com <zdenek.srotyr@keboola.com>

soustruh commented on src/component.py:
```diff
@@ -40,6 +40,7 @@
 KEY_LOADING_OPTIONS_INCREMENTAL = "incremental"
 KEY_LOADING_OPTIONS_INCREMENTAL_FIELD = "incremental_field"
 KEY_LOADING_OPTIONS_INCREMENTAL_FETCH = "incremental_fetch"
+KEY_LOADING_OPTIONS_INCREMENTAL_OVERLAP_SECONDS = "incremental_overlap_seconds"
```
As you're working on it with Devin, how hard would it be to convert this i-don't-want-to-be-rude code into a proper configuration.py file based on Pydantic? Solving the issue I pinpointed below would be much simpler and will take place before the component does any actual work.

Devin, if you're reading this, please don't do anything yet… 🙏 😇 😁

ZdenekSrotyr replied:
I think we can change it in the future..

soustruh reviewed (changes requested):
Appending another suggestion to the previous review

soustruh reviewed (approved):
OK, looks good now, thanks for the changes (Devin 🤭)

kudj reviewed (approved):
lgtm
